### PR TITLE
[FSDP] Option to keep grads in lower prec

### PR DESCRIPTION
**Pull request by rohan-varma**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #85134

Differential Revision: [D39565189](https://our.internmc.facebook.com/intern/diff/D39565189)

Rehash of a similar PR from a month ago that got stale. Adds a config to FSDP MP so that gradients can be kept in lower precision, to support optimizers such as AnyPrecisionOptimizer which would like to keep grads in bf16. 

To do this, for sharded cases, we cannot simply omit the cast back to the full precision param dtype, otherwise when setting `p.grad = p._saved_grad_shard` in finalize_params, autograd will throw an error indicating that the grad dtype should match the param dtype when it is being set.

As a workaround, we re-cast after setting this. Although, this means that for cases that use gradient accumulation, p._saved_grad_shard will be of the reduced dtype because it is set to p.grad in `_prep_grad_for_backward`. As a result, add a check + recast here as well.

### CODE REVIEW

**Comment by rohan-varma** [comment]
@rohan-varma has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39565189).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -179,6 +179,7 @@ class MixedPrecision:
         are checkpointed in their full precision (and then restored back to
         to their reduced precision) as expected. Note that this checkpoint
         support is currently limited to ``StateDictType.FULL_STATE_DICT``.
+        ``keep_casted_gradients``: Keep casted gradients.
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:182]: small nit (grammar): It appears the past tense conjugation of "cast" is "cast", not "casted".
- https://writingexplained.org/cast-vs-casted (search for "What does casted mean?")
- https://pasttenses.com/cast-past-tense

I am indifferent to `keep_cast_gradients` (emphasizing we are keeping the gradients, which happen to have been cast) versus `keep_gradients_cast` (emphasizing that we are keeping them cast).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -3403,6 +3423,10 @@ def _finalize_params(fsdp_module: FullyShardedDataParallel) -> None:
                         # lands. If it was not called, there is no new gradient to accumulate
                         if p._post_backward_called:
                             p.grad = p._saved_grad_shard
+                            if fsdp_module._mixed_precision_keep_low_precision_grads():
+                                p.grad.data = p.grad.data.to(
```
Comment (awgu) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:3427]: IIUC, we _need_ this `.data` to bypass the autograd check that the parameter dtype matches the gradient dtype?

Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:3427]: nit: I think this should avoid one `aten.detach` op and work the same.
```suggestion
                                p.grad.data = p.grad.to(
```

**Review by awgu — COMMENTED** [review]
Is there a way to design this to avoid casting to FP32 and casting back to the low precision?

Perhaps, Core team can expose a way to bypass the requirement for the parameter and gradient to have the same dtype since the `AnyPrecisionOptimizer` is a real, tame use case.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:182]: Okay I have been saying `keep_casted_gradients`, `keep_cast_gradients`, `keep_gradients_cast`, and `keep_gradients_casted` in my head, and none of them particularly sound correct. Feel free to go with whichever you prefer :)

Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:182]: Interesting, yeah keep_cast_gradients sounds kinda strange to me as well.

**Comment by rohan-varma** [comment]
> Is there a way to design this to avoid casting to FP32 and casting back to the low precision?
> 
> Perhaps, Core team can expose a way to bypass the requirement for the parameter and gradient to have the same dtype since the `AnyPrecisionOptimizer` is a real, tame use case.

Yeah, we would need some support / bypass capability from the core side to design this such that we don't upcast and downcast again. 

The issue can be reproduced as such: 

```
>>> import torch
>>> z = torch.nn.Linear(3,3)
>>> z
Linear(in_features=3, out_features=3, bias=True)
>>> inp = torch.randn(3, 3)
>>> out = z(inp).sum()
>>> out
tensor(-3.4956, grad_fn=<SumBackward0>)
>>> out.backward(
... )
>>> z.weight.grad
tensor([[ 2.9189, -1.5294,  0.9937],
        [ 2.9189, -1.5294,  0.9937],
        [ 2.9189, -1.5294,  0.9937]])
>>> z.weight.grad = z.weight.grad.to(torch.bfloat16)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/torch/_tensor.py", line 1100, in grad
    self._grad = new_grad
RuntimeError: assigned grad has data of a different type
>>> bf16_grad = torch.ones_like(z.weight.grad).to(torch.bfloat16)
>>> z.weight.grad = bf16_grad
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Library/Frameworks/Python.framework/Versions/3.7/lib/python3.7/site-packages/torch/_tensor.py", line 1100, in grad
    self._grad = new_grad
RuntimeError: assigned grad has data of a different type
```

In FSDP, this would occur when we assign `p.grad = p._saved_grad_shard` and if `_saved_grad_shard` was the reduced dtype. As a result, we upcast and then downcast with the workaround `p.grad.data = p.grad.to(...)` to switch the type. It's pretty hacky, but I can't see any other way to bypass the "assigned grad has data of a different type" issue. 

Will sync up with the core team to see if we can enhance this, but my thinking is that should not block us unless you disagree.

**Comment by awgu** [comment]
> Will sync up with the core team to see if we can enhance this, but my thinking is that should not block us unless you disagree.

Got it. This makes sense to me. I think if we are already seeing good results with the upcast + downcast then, this is good to land.

**Review by awgu — APPROVED** [review]
LGTM!

**Comment by rohan-varma** [comment]
@rohan-varma has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D39565189).

**Comment by ZainRizvi** [comment]
@pytorchmergebot revert -c nosignal -m "broke trunk, failing the tests test_grads_reduced_precision (__main__.TestFSDPMixedPrecisionUnsharded)"

This PR caused the test test_grads_reduced_precision (__main__.TestFSDPMixedPrecisionUnsharded) to start failing. Please see this link for more detailed logs:
https://hud.pytorch.org/pytorch/pytorch/commit/607eccb13ca586f775fb09daeb728a4b4e30ebdd

**Comment by ZainRizvi** [comment]
@pytorchmergebot revert -c nosignal -m "broke trunk, failing the tests test_grads_reduced_precision (main.TestFSDPMixedPrecisionUnsharded)"

This PR caused the test test_grads_reduced_precision (__main__.TestFSDPMixedPrecisionUnsharded) to start failing. Please see this link for more detailed logs:
https://hud.pytorch.org/pytorch/pytorch/commit/607eccb13ca586f775fb09daeb728a4b4e30ebdd

**Comment by rohan-varma** [comment]
Actually not sure if reopen + push to the reverted PR is the recommended practice. Created reland https://github.com/pytorch/pytorch/pull/85223

**Pull request closed without merging**